File: 0295-find-median-from-data-stream/0295-find-median-from-data-stream.cpp
class MedianFinder {
private:
    priority_queue<int> lo;
    priority_queue<int, vector<int>, greater<int>> hi;
public:
    MedianFinder() {
        // leave it as it is.
    }
    
    void addNum(int num) {
        lo.push(num);
        hi.push(lo.top());
        lo.pop();
        if(lo.size() < hi.size()){
            lo.push(hi.top());
            hi.pop();
        }
    }
    
    double findMedian() {
        double ans = 1.00 * lo.top();
        if(lo.size() == hi.size()){
            ans += hi.top();
            return ans * 0.5;
        }
        return ans;
    }
};

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder* obj = new MedianFinder();
 * obj->addNum(num);
 * double param_2 = obj->findMedian();
 */

File: 0295-find-median-from-data-stream/0295-find-median-from-data-stream.py
class MedianFinder:
    def __init__(self):
        self.left = [] # Max heap
        self.right = [] # Min heap

    def addNum(self, num: int) -> None:
        heapq.heappush(self.left, -1 * num)
        n = -1 * heapq.heappop(self.left)
        heapq.heappush(self.right, n)
        if len(self.right) > len(self.left):
            n = heapq.heappop(self.right)
            heapq.heappush(self.left, -1 * n)
        

    def findMedian(self) -> float:
        if (len(self.left) + len(self.right)) % 2 == 0:
            return (-1 * self.left[0] + self.right[0]) / 2.0
        return -1 * self.left[0]


# Your MedianFinder object will be instantiated and called as such:
# obj = MedianFinder()
# obj.addNum(num)
# param_2 = obj.findMedian()